sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuParseUrl.scala

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.rapids
+
+import ai.rapids.cudf.ColumnVector
+import com.nvidia.spark.rapids._
+import com.nvidia.spark.rapids.Arm._
+import com.nvidia.spark.rapids.RapidsPluginImplicits._
+import com.nvidia.spark.rapids.jni.ParseURI
+import com.nvidia.spark.rapids.shims.ShimExpression
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.vectorized.ColumnarBatch
+import org.apache.spark.unsafe.types.UTF8String
+
+object GpuParseUrl {
+  val HOST = "HOST"
+  val PATH = "PATH"
+  val QUERY = "QUERY"
+  val REF = "REF"
+  val PROTOCOL = "PROTOCOL"
+  val FILE = "FILE"
+  val AUTHORITY = "AUTHORITY"
+  val USERINFO = "USERINFO"
+
+  def isSupportedPart(part: String): Boolean = {
+    part match {
+      case PROTOCOL | HOST =>
+        true
+      case _ =>
+        false
+    }
+  }
+}
+
+case class GpuParseUrl(children: Seq[Expression]) 
+  extends GpuExpression with ShimExpression with ExpectsInputTypes {
+
+  override def nullable: Boolean = true
+  override def inputTypes: Seq[DataType] = Seq.fill(children.size)(StringType)
+  override def dataType: DataType = StringType
+  override def prettyName: String = "parse_url"
+
+  import GpuParseUrl._
+
+  def doColumnar(url: GpuColumnVector, partToExtract: GpuScalar): ColumnVector = {
+    val part = partToExtract.getValue.asInstanceOf[UTF8String].toString
+    part match {
+      case PROTOCOL =>
+        ParseURI.parseURIProtocol(url.getBase)
+      case HOST =>
+        ParseURI.parseURIHost(url.getBase)
+      case PATH | QUERY | REF | FILE | AUTHORITY | USERINFO =>
+        throw new UnsupportedOperationException(s"$this is not supported partToExtract=$part. " +
+            s"Only PROTOCOL and HOST are supported")
+      case _ =>
+        throw new IllegalArgumentException(s"Invalid partToExtract: $partToExtract")
+    }
+  }
+
+  def doColumnar(url: GpuColumnVector, partToExtract: GpuScalar, key: GpuScalar): ColumnVector = {
+    val part = partToExtract.getValue.asInstanceOf[UTF8String].toString
+    if (part != QUERY) {
+      // return a null columnvector
+      return ColumnVector.fromStrings(null, null)
+    }
+    throw new UnsupportedOperationException(s"$this only supports partToExtract = PROTOCOL")
```

Comment:
The message seems to be vague here. We say we are supporting PROTOCOL and HOST so here needs more explanation.